Implement multi-community (fixes #818, fixes #5340)

Nutomic opened the pull request.

dullbananas commented on crates/db_schema/src/impls/multi_community.rs:
```diff
@@ -0,0 +1,135 @@
+      ))
+      .into_boxed();
+
+    let (multi, entries) = match params {
+      ReadParams::Name {
+        user_name,
+        multi_name,
+      } => query
+        .filter(person::name.eq(user_name))
+        .filter(multi_community::name.eq(multi_name)),
+      ReadParams::Id(id) => query.filter(multi_community::id.eq(id)),
+      ReadParams::ApId(ap_id) => query.filter(multi_community::ap_id.eq(ap_id)),
+    }
+    .first(conn)
+    .await?;
```
less confusing:
```suggestion
    query = match params {
      ReadParams::Name {
        user_name,
        multi_name,
      } => query
        .filter(person::name.eq(user_name))
        .filter(multi_community::name.eq(multi_name)),
      ReadParams::Id(id) => query.filter(multi_community::id.eq(id)),
      ReadParams::ApId(ap_id) => query.filter(multi_community::ap_id.eq(ap_id)),
    };
    let (multi, entries) = query.first(conn)
      .await?;
```

Nutomic replied:
Done

dullbananas commented on crates/db_views/post/src/impls.rs:
```diff
@@ -344,8 +362,21 @@ impl<'a> PostQuery<'a> {
+        query = query.filter(post::community_id.eq(id));
+      }
+      Some(CommunityOrMulti::MultiCommunityId(id)) => {
+        let multi = MultiCommunity::read(pool, ReadParams::Id(id)).await?;
+        query = query.filter(post::community_id.eq_any(multi.entries));
```
```suggestion
        let communities = multi_community_entry::table.filter(multi_community_entry::multi_community_id.eq(id));
        query = query.filter(post::community_id.eq_any(communities));
```

This subquery can be evaluated once for the whole query, so it shouldn't be a performance problem, and is better than a separate query

https://www.postgresql.org/docs/17/using-explain.html#:~:text=In%20the%20example%20above,operator%20be%20amenable%20to%20hashing

Nutomic replied:
This is not compiling and throws a cryptic diesel error:
```

error[E0277]: the trait bound `SelectStatement<FromClause<lemmy_db_schema_file::schema::multi_community_entry::table>, query_builder::select_clause::DefaultSelectClause<FromClause<lemmy_db_schema_file::schema::multi_community_entry::table>>, query_builder::distinct_clause::NoDistinctClause, query_builder::where_clause::WhereClause<diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<multi_community_id, diesel::expression::bound::Bound<Integer, MultiCommunityId>>>>>: AsInExpression<Integer>` is not satisfied
--> crates/db_views/post/src/impls.rs:380:56
|
380 |         query = query.filter(post::community_id.eq_any(communities));
|                                                 ------ ^^^^^^^^^^^ the trait `AsInExpression<Integer>` is not implemented for `SelectStatement<FromClause<table>, DefaultSelectClause<FromClause<table>>, NoDistinctClause, WhereClause<Grouped<...>>>`
|                                                 |
|                                                 required by a bound introduced by this call
|
```

Maybe its missing the join to multi_community_entry table. I wasnt able to get the join working inside this match, so it would probably have to be done in the general `joins()` function. That seems like a bad idea because frontpage queries would run into the join limit.

dullbananas replied:
Just add `.select(multi_community_entry::community_id)`

Nutomic replied:
Ah didnt know that this works.

dullbananas commented on crates/utils/src/utils/validation.rs:
```diff
@@ -126,10 +126,10 @@ fn has_3_permitted_display_chars(name: &str) -> bool {
   let check = !name.starts_with('@')
     && !name.starts_with(FORBIDDEN_DISPLAY_CHARS)
-    && name.chars().count() <= actor_name_max_length
+    && name.chars().count() as i32 <= actor_name_max_length
```
use `try_from` or `try_into` instead of `as`

Nutomic replied:
Will be fixed with https://github.com/LemmyNet/lemmy/pull/5688

Nutomic commented on crates/api_crud/src/multi_community/create.rs:
```diff
@@ -0,0 +1,35 @@
+  data: Json<CreateMultiCommunity>,
+  context: Data<LemmyContext>,
+  local_user_view: LocalUserView,
+) -> LemmyResult<Json<MultiCommunity>> {
```
How many items should we allow to have acceptable performance?

dullbananas replied:
`MAX_API_PARAM_ELEMENTS` should be used

Nutomic replied:
Which is defined as 10k, thats a lot. Anyway I set that for now and we can change it later.

Nutomic commented on crates/api_crud/src/multi_community/update.rs:
```diff
@@ -0,0 +1,40 @@
+    return Err(LemmyErrorType::NotFound.into());
+  }
+  // Disallow removed/deleted communities
+  try_join_all(data.communities.iter().map(|id| async {
+    let c = Community::read(&mut context.pool(), *id).await?;
+    if c.removed || c.deleted || c.visibility != CommunityVisibility::Public {
+      return Err(LemmyErrorType::NotFound.into());
+    }
+    Ok::<_, LemmyError>(c)
+  }))
+  .await?;
```
You mean it should be possible to add removed or deleted communities to a  multi-comm? post_view.rs already has its own filters for removed/deleted community.

dullbananas replied:
A filter when getting a list of entries is more convenient because otherwise users would need to manually remove entries for deleted/removed communities before they can add another community.

Now that I think of it more, it's unclear whether or not that filter would really make this check in the update API unnecessary.

Nutomic replied:
Obsolete now that this was replaced with separate add/delete entry endpoints.

Nutomic commented on crates/apub/src/collections/multi_community.rs:
```diff
@@ -0,0 +1,119 @@
+    )
+    .await
+    .into_iter()
+    .flat_map(|c: LemmyResult<CommunityId>| c.ok())
```
Added log for errors.

dullbananas replied:
That doesn't fix possibly misleading `Ok` return values for `from_json`. It's okay for this PR because I just saw that ignoring errors is already done for other apub collections, but I will open an issue.

Nutomic commented on crates/db_schema/src/impls/multi_community.rs:
```diff
@@ -0,0 +1,135 @@
+  ) -> Result<MultiCommunityView, Error> {
+    let conn = &mut get_conn(pool).await?;
+    let query = multi_community_entry::table
+      .left_join(multi_community::table.left_join(person::table))
+      .filter(multi_community::id.is_not_null())
```
Is there a better way to get the `multi_community` and all associated `multi_community_entry`s?

dullbananas replied:
I just realized that the current implementation appears to only do that correctly for a non-empty multi-community, and my suggested change doesn't fix it. Ignore that suggested change.

Fix:
* Add a test that calls this function on an empty multi-community
* Correctly handle the value returned by arrag_agg when there's no items (it returns null instead of an empty array)
* Swap multi_community_entry and multi_community in the join order, so a multi community can be selected even if there's no associated entries (also, with this change, the not-null filter no longer makes a difference, so it can be removed)

dullbananas replied:
Add a comment about the `[null]` result, to prevent confusion about what difference the `array_remove` call makes.

Also it seems like the `null` in that array corresponds to the `null` value that exists in the single row returned by the left join.

Nutomic replied:
Thats true. Added comment.

Nutomic commented on crates/apub/src/api/list_posts.rs:
```diff
@@ -87,6 +91,13 @@ pub async fn list_posts(
+    (None, Some(id)) => Some(CommunityOrMulti::MultiCommunityId(id)),
+    (Some(_), Some(_)) => return Err(LemmyErrorType::InvalidCommunityId.into()),
+    (None, None) => None,
+  };
```
Any suggestions how the api and Postquery para should look like? At the moment it feels a bit awkward.

Nutomic replied:
I removed the CommunityOrMulti enum, but queries are already working fine and are better this way.

Nutomic commented:
Right the removed/deleted filters for read functions were missing, added those now. Also changed the update endpoint to filter out removed/deleted communities instead of throwing error.

Nutomic commented on crates/db_views/community/src/multi_community.rs:
```diff
@@ -0,0 +1,151 @@
+      .left_join(multi_community_entry::table.inner_join(community::table))
+      .group_by(multi_community::id)
+      .filter(not(community::removed))
+      .filter(not(community::deleted))
```
Somehow this is not working and always returns not_found. I already tried adding `.nullable()` but same. Any other ideas?

dullbananas replied:
Use filter(community::removed.or(community::deleted).is_distinct_from(true)) so it works even if community is null.

dessalines replied:
Its because you're trying to take a single multi_community, and join it to a list of things, which isn't how SQL works.

IMO you don't need this view at all. In the `multi_community/get` route, fetch the multi_community, and the list of `communities` separately. There's no need to join the multi_community to each community fetched. IE `MultiCommunity::list_communities(multi_community_id)`

dessalines commented on crates/db_schema/src/impls/multi_community.rs:
```diff
@@ -0,0 +1,84 @@
+      .await
+  }
+
+  pub async fn upsert(
```
You probably won't need this after switching to atomic updates.

Nutomic replied:
This is needed for federation which uses a full list of communities, so it cant be updated one by one.

dessalines commented on crates/db_schema/src/impls/multi_community.rs:
```diff
@@ -0,0 +1,84 @@
+use diesel_async::{scoped_futures::ScopedFutureExt, AsyncConnection, RunQueryDsl};
+use lemmy_db_schema_file::schema::{multi_community, multi_community_entry};
+
+impl MultiCommunity {
+  pub async fn create(
```
Would be much better to just implement Crud. Check out `crates/db_schema/src/impls/oauth_provider.rs` for a close example, it should have everything you need.

Nutomic replied:
Had that before but it didnt work with transactions (because transactions pass DbConn param but Crud only allows DbPool).

dessalines commented on crates/db_schema/src/impls/multi_community.rs:
```diff
@@ -0,0 +1,84 @@
+      .await
+  }
+
+  pub async fn update(
```
You won't need this either.

dessalines commented on crates/db_schema/src/impls/multi_community.rs:
```diff
@@ -0,0 +1,84 @@
+    Ok(())
+  }
+
+  pub async fn list(pool: &mut DbPool<'_>, owner_id: Option<PersonId>) -> Result<Vec<Self>, Error> {
```
Needs to be a lemmyerror, I just did a big refactor to get all these into lemmy errors. Check crates/db_schema/src/impls/oauth_provider.rs

dessalines commented on crates/db_views/community/src/lib.rs:
```diff
@@ -14,6 +18,8 @@ use {
 
 #[cfg(feature = "full")]
 pub mod impls;
+#[cfg(feature = "full")]
+pub mod multi_community;
```
This view isn't necessary.

dessalines commented on crates/db_views/community/src/lib.rs:
```diff
@@ -43,3 +49,17 @@ pub struct CommunityView {
   pub post_tags: TagsView,
 }
+
+#[derive(Debug, PartialEq, Serialize, Deserialize, Clone)]
+#[cfg_attr(feature = "full", derive(TS))]
+#[cfg_attr(feature = "full", ts(export))]
+pub struct MultiCommunityView {
+  pub multi: MultiCommunity,
+  pub entries: Vec<CommunityId>,
+}
```
Delete. Fetching the multi_community, and the list of communities it contains, need to be different fetches anyway.

dessalines commented on crates/db_schema/src/impls/multi_community.rs:
```diff
@@ -0,0 +1,84 @@
+      query = query.filter(multi_community::owner_id.eq(owner_id));
+    }
+    query.get_results(conn).await
+  }
```
This also needs a list of the communities for that multi community:

```rust
fn list_communities(multi_community_id: MultiCommunityId) -> LemmyResult<Vec<Community>> {
  ...
  multi_community_entry::table
  .inner_join(multi_community::table)
  .inner_join(community::table)
  .filter(multi_community::id.eq(multi_community_id)
  .select(Community::as_select())
```

Nutomic replied:
So instead of the views which include all the data and are already working, you want to go back to separate queries? Seems pointless and makes worse performance and more complicated code.

dessalines replied:
Views should never contain vectors, I think this would be the very first view that does. Look at every other db_view `lib.rs`, you'll see they're all doing simple diesel embed columns.

The API response for a GET multi-community should be : 

```rust
multi_community: MultiCommunity,
entries: Vec<Community (or a communityview)>
```

So its easy enough to just fetch these separately, and no need to use group bys or anything. 

You *could* create a view for `MultiCommunityEntry`, where each entry embeds the MultiCommunity, but I don't think its worth it, especially since you don't need to query these in complex ways, in the same way we need to with PostQuery or CommunityQuery. You only ever want the entries for a single `multi_community_id`. But if you wanted to, then you could, it would just contain a lot of pointless extra data.

Nutomic replied:
Makes sense that it should return the full community details for Get call. Implemented that via CommunityQuery. Also changed resolve_object to return simply MultiCommunity in order to get rid of MultiCommunityView entirely. There is still MultiCommunityApub in use (maybe some of these structs need to be renamed).

dessalines replied:
Cool. Its pry not good idea to keep the MultiCommunityView just for apub, you'll run into a lot of complications I imagine.

Nutomic replied:
Its not really a view so I renamed and moved it to db_schema.

dessalines commented on crates/db_views/post/src/impls.rs:
```diff
@@ -234,12 +244,30 @@ impl PostView {
 }
 
+#[derive(Clone)]
+pub enum CommunityOrMulti {
```
See above, its not necessary.
just add `multi_community_ids: Option<Vec<CommunityId>>` as a field to postquery, and use `.eq_any`

dessalines commented on crates/routes/src/utils/scheduled_tasks.rs:
```diff
@@ -557,6 +566,33 @@ async fn build_update_instance_form(
 }
 
+/// Updates all multi-communities from remote users
+async fn update_multi_communities(context: &Data<LemmyContext>) -> LemmyResult<()> {
```
I don't really understand why this is necessary. 

Shouldn't they just get initially fetched on demand just like `Community`, and updates pushed after subscribing?

dessalines reviewed (commented):
I don't fully understand the apub stuff, but I feel like subscribing to them should work similar to communities, and that we keep to the push paradigm of apub.

Nutomic commented on crates/api_common/src/post.rs:
```diff
@@ -122,6 +131,8 @@ pub struct GetPosts {
   pub page_back: Option<bool>,
   #[cfg_attr(feature = "full", ts(optional))]
   pub limit: Option<i64>,
+  #[cfg_attr(feature = "full", ts(optional))]
+  pub multi_community_id: Option<MultiCommunityId>,
```
Done

Nutomic commented on migrations/2025-05-13-103308_multi-community/up.sql:
```diff
@@ -0,0 +1,19 @@
+    PRIMARY KEY (multi_community_id, community_id)
+);
+
+ALTER TABLE local_site
+    ADD COLUMN suggested_communities int REFERENCES multi_community ON UPDATE CASCADE ON DELETE CASCADE;
```
Would be too complicated to split these changes up now.

Nutomic commented on crates/api_common/src/community.rs:
```diff
@@ -362,3 +362,33 @@ pub struct ApproveCommunityPendingFollower {
+#[cfg_attr(feature = "full", ts(export))]
+pub struct UpdateMultiCommunity {
+  pub id: MultiCommunityId,
+  pub communities: Vec<CommunityId>,
```
Done

Nutomic commented:
About federation, sending update activities for changes to multi-community is not that simple. The way its implemented now multi-comms are associated to a person, so it would mean following the person and also getting a lot of unrelated activities for new posts and comments. Piefed did it by implementing a new actor type (similar to community or person), which means a lot of overhead for something relatively minor. 

Another possibility would be to associate multi-comms with the instance instead of a person. Its not clear who would send the follow in that case, plus it means that updates to all multi-commns get sent. It could work by sending updates for all local multi-comms to all known instances, and the receiving side ignores those which havent been fetched already.

dessalines commented on crates/api_crud/src/multi_community/update.rs:
```diff
@@ -0,0 +1,38 @@
+  local_user_view: LocalUserView,
+) -> LemmyResult<Json<SuccessResponse>> {
+  // check that owner is correct
+  let read = MultiCommunityView::read(&mut context.pool(), ReadParams::Id(data.id)).await?;
+  if read.multi.creator_id != local_user_view.person.id {
+    return Err(LemmyErrorType::MultiCommunityUpdateWrongUser.into());
+  }
```
You are doing this in multiple places, so might as well move it to a common function.

Nutomic replied:
Done

dessalines commented on crates/api_crud/src/multi_community/delete_entry.rs:
```diff
@@ -0,0 +1,27 @@
+  context: Data<LemmyContext>,
+  local_user_view: LocalUserView,
+) -> LemmyResult<Json<SuccessResponse>> {
+  // check that owner is correct
```
Here's the other one.

dessalines commented on crates/api_crud/src/multi_community/create.rs:
```diff
@@ -0,0 +1,40 @@
+    name: data.name.clone(),
+    title: data.title.clone(),
+    description: data.description.clone(),
+    ap_id: Url::parse(&format!(
+      "{}/m/{}",
+      &local_user_view.person.ap_id, &data.name
+    ))?
+    .into(),
```
Don't we have some trait / helper functions to create these somewhere?

Okay I see that both `Community` and `Person` impls have a `local_url` function. It'd be a good idea to make this function a trait, and have community, person, and multi_community implement it.

Nutomic replied:
Those are called from multiple places, but here its only used in a single place. Would be unnecessary to add a separate method.

dessalines commented on crates/api_crud/src/multi_community/create_entry.rs:
```diff
@@ -0,0 +1,34 @@
+  context: Data<LemmyContext>,
+  local_user_view: LocalUserView,
+) -> LemmyResult<Json<SuccessResponse>> {
+  // check that owner is correct
```
Here's another one.

dessalines commented on crates/api_common/src/community.rs:
```diff
@@ -362,3 +362,57 @@ pub struct ApproveCommunityPendingFollower {
+#[cfg_attr(feature = "full", ts(export))]
+pub struct ListMultiCommunities {
+  #[cfg_attr(feature = "full", ts(optional))]
+  pub owner_id: Option<PersonId>,
```
`creator_id`

dessalines commented on crates/db_schema/src/impls/multi_community.rs:
```diff
@@ -0,0 +1,127 @@
+    )
+  }
+
+  pub async fn create_entry(
+    pool: &mut DbPool<'_>,
+    id: MultiCommunityId,
+    new_community: CommunityId,
+  ) -> LemmyResult<()> {
```
This should probably be on the crud impl of the `MultiCommunityEntry` struct, and it should return `LemmyResult<Self>`

dessalines reviewed (commented):
> Piefed did it by implementing a new actor type (similar to community or person), which means a lot of overhead for something relatively minor.

That's unavoidable I think. If we let people follow multicommunities, then that is a specific actor type similar to communities. Its also distinct from following a person (even if we did support that) because you might want to follow a specific multi-community created by a person, but not all of the ones they created. There's probably no way around adding `MultiCommunity` as a followable actor type.

dessalines commented on crates/api_crud/src/multi_community/get.rs:
```diff
@@ -0,0 +1,25 @@
+) -> LemmyResult<Json<GetMultiCommunityResponse>> {
+  let local_site = SiteView::read_local(&mut context.pool()).await?;
+  let multi = MultiCommunity::read(&mut context.pool(), data.id).await?;
+  let entries = CommunityQuery {
+    multi_community_id: Some(multi.id),
+    ..Default::default()
+  }
+  .list(&local_site.site, &mut context.pool())
+  .await?;
```
Nice, this is perfect.

dessalines commented:
> Follow implies that communities from followed multi-comms would show up in the subscribed feed which is probably not what users want.

Yep this is a unique type of follow that doesn't affect front page feeds, but would allow updates to that multicommunity to be pushed to instances with users that follow it, and allow local representations across different servers. Seems unavoidable.

dessalines reviewed (commented):
Once you think this is ready ping me, (or mark / unmark as draft). I have some very minor code changes I'd like to make, but I don't want to touch it until you're ready.

Nutomic commented:
Alright made a lot more changes which should address all the feedback and implement all the functionality. Still need to expand the api test. Please have a look.

dullbananas commented on crates/db_schema/src/impls/multi_community.rs:
```diff
@@ -1,84 +1,663 @@
+      .into_boxed();
+    if is_removed_from_multi {
+      // Remove: only in case it was a multi-comm follow (not manually by user)
+      query = query.filter(community_actions::is_multi_community_follow.is_not_null());
```
```suggestion
      query = query.filter(community_actions::is_multi_community_follow.is_distinct_from(false));
```

dullbananas commented on crates/db_schema/src/impls/multi_community.rs:
```diff
@@ -1,84 +1,663 @@
+    Ok(())
+  }
+
+  pub async fn update_local_follows(
```
```suggestion
  pub async fn update_follows_to_local_community(
```

dullbananas commented on crates/db_schema/src/impls/multi_community.rs:
```diff
@@ -1,84 +1,663 @@
+      .distinct()
+      .load(conn)
       .await
+      .with_lemmy_type(LemmyErrorType::NotFound)
```
```suggestion
      .optional()?
      .ok_or(LemmyErrorType::NotFound.into())
```

dullbananas commented on crates/db_schema/src/impls/multi_community.rs:
```diff
@@ -1,84 +1,663 @@
+      .await?;
+
+    // get all local users who follow the multi-comm
+    let local_followers: Vec<Person> = multi_community_follow::table
+      .left_join(person::table)
+      .inner_join(multi_community::table)
+      .filter(person::id.is_not_null())
+      .filter(person::local)
+      .select(person::all_columns.nullable().assume_not_null())
```
```suggestion
    let local_followers: Vec<Person> = multi_community_follow::table
      .inner_join(person::table)
      .inner_join(multi_community::table)
      .filter(person::local)
      .select(person::all_columns)
```

Nutomic replied:
Done

dullbananas commented on migrations/2025-05-28-103308_multi-community/up.sql:
Missing indexes for filteres.

dessalines commented on crates/api/src/community/multi_community_follow.rs:
```diff
@@ -0,0 +1,91 @@
+    follow_state,
+  };
+
+  if data.follow {
+    MultiCommunity::follow(&mut context.pool(), &form).await?;
+
+    // Get all communities which are part of the multi and not yet followed by current user
+    let to_follow: Vec<_> = communities
+      .into_iter()
+      .filter(|c| {
+        let actions = c.community_actions.clone().unwrap_or_default();
+        actions.followed.is_none()
+      })
+      .map(|c| c.community)
+      .collect();
+
+    // Then follow them
+    community_follow_many(&local_user_view.person, &to_follow, &context).await?;
```
The way it works on reddit, just because you follow a multi-community does not mean you've subscribed to every community inside it.

I spose since you can't guarantee that these communities have local subscribers, then this is unavoidable. Just wanted to point out this difference.

The only other way around this I can think of, is to keep a fake user just for remote subscribes.

Nutomic replied:
Good idea, that will make the logic a lot simpler. I should have thought of that before writing all this code.

Nutomic replied:
Implemented the schema for this now.

dessalines commented on crates/api/src/community/multi_community_follow.rs:
```diff
@@ -0,0 +1,91 @@
+  } else {
+    MultiCommunity::unfollow(&mut context.pool(), person_id, multi_community_id).await?;
+
+    // Unfollow all communities which were followed as part of multi-comm
+    // (is_multi_community_follow=true)
+    // TODO: what if a user follows more than one multi-comm
+    // containing the same community? then it would get wrongly removed here. so it needs a
+    // separate db query to check that.
+    let to_unfollow: Vec<_> = communities
```
I'd say it's probably safer to just not unfollow communities. You can't know that they want to unfollow one of the contained multicommunities just because they unfollow the multi.

dessalines commented on crates/api_common/src/community.rs:
```diff
@@ -362,3 +362,75 @@ pub struct ApproveCommunityPendingFollower {
+#[cfg_attr(feature = "full", ts(export))]
+pub struct GetMultiCommunityResponse {
+  pub multi: MultiCommunity,
+  pub entries: Vec<CommunityView>,
```
I think i'd prefer this field be called `communities`

dessalines commented on crates/api_common/src/post.rs:
```diff
@@ -96,6 +105,7 @@ pub struct GetPosts {
   pub community_name: Option<String>,
   #[cfg_attr(feature = "full", ts(optional))]
+  pub multi_community_id: Option<MultiCommunityId>,
   pub show_hidden: Option<bool>,
```
Needs ts_optional

dessalines commented on crates/api_common/src/utils.rs:
```diff
@@ -1000,6 +1008,56 @@ pub fn send_webmention(post: Post, community: &Community) {
   };
 }
 
+pub fn community_follower_state(community: &Community) -> CommunityFollowerState {
+  if community.local {
+    // Local follow is accepted immediately
+    CommunityFollowerState::Accepted
+  } else if community.visibility == CommunityVisibility::Private {
+    // Private communities require manual approval
```
Shouldn't the private check go first? Otherwise local private ones immediately get accepted. I forget how federation works with these.

Nutomic replied:
True

dessalines commented on crates/api_common/src/utils.rs:
```diff
@@ -1000,6 +1008,56 @@ pub fn send_webmention(post: Post, community: &Community) {
+) -> LemmyResult<()> {
+  for community in to_follow {
+    let state = community_follower_state(community);
+    let mut form = CommunityFollowerForm::new(community.id, person.id, state);
+    form.is_multi_community_follow = Some(true);
+    CommunityActions::follow(&mut context.pool(), &form).await?;
```
No need for the mut, just do 

```rust
CommunityFollowerForm {
  is_multi_community: Some(true),
  ..CommunityFollowerForm::new(community.id, person.id, state)
}
```

I'm not sure this DB field makes sense tho. I'll skip below.

dessalines commented on migrations/2025-06-04-103308_multi-community/up.sql:
```diff
@@ -0,0 +1,49 @@
+ALTER TYPE listing_type_enum
+    ADD VALUE 'Suggested';
+
+ALTER TABLE community_actions
+    ADD COLUMN is_multi_community_follow bool;
```
Okay I see this now. But the scary thing here is that this table has a primary key of `person_id, community_id`.

If a community is part of a multi-sub, then you do an ordinary subscribe, it needs to get *upgraded* by setting `is_multi_community_follow` to false. 

If you unfollow, it should only unfollow rows where `is_multi_community` is false. 

This is a bit complicated to try to put both of these in one table, and really it could be solved easier by having a fake multi_community follower user handle all these, rather than adding this column.

dullbananas replied:
Another solution is to only add a `community_actions.multi_community_follows` column, which counts the number of multi-communities that both contain this community and are followed by this user. This allows an approach that's very similar to reference counting. Following a multi-community increments, and unfollowing decrements. 0 must be stored as null because of how uplete works.

Subscribed-only filters would become `community_actions::followed.is_not_null().or(community_actions::multi_community_follows.is_not_null())`.

Nutomic replied:
Im implementing a dummy user to follow these communities and getting rid of this column.

dullbananas commented on migrations/2025-06-04-103308_multi-community/up.sql:
```diff
@@ -0,0 +1,85 @@
+ALTER TYPE listing_type_enum
+    ADD VALUE 'Suggested';
+
+CREATE INDEX idx_multi_community_read_from_name ON multi_community (local, deleted, name);
```
```suggestion
CREATE INDEX idx_multi_community_read_from_name ON multi_community (name) WHERE local AND NOT deleted;
```

dullbananas commented on migrations/2025-06-04-103308_multi-community/up.sql:
```diff
@@ -0,0 +1,85 @@
+
+CREATE TABLE multi_community_entry (
+    multi_community_id int NOT NULL REFERENCES multi_community ON UPDATE CASCADE ON DELETE CASCADE,
+    community_id int NOT NULL REFERENCES community ON UPDATE CASCADE ON DELETE CASCADE,
```
`community_id` needs to be indexed for the `community_used_in_multiple` function.

Nutomic commented on crates/db_schema/src/test_data.rs:
```diff
@@ -0,0 +1,42 @@
+    Site::delete(pool, self.site.id).await?;
+    Ok(())
+  }
+}
```
The same or similar logic was used in lots of different tests, so I made this helper to deduplicate things a bit.

dessalines commented on migrations/2025-06-04-103308_multi-community/up.sql:
```diff
@@ -0,0 +1,85 @@
+    multi_community_id int NOT NULL REFERENCES multi_community ON UPDATE CASCADE ON DELETE CASCADE,
+    person_id int NOT NULL REFERENCES person ON UPDATE CASCADE ON DELETE CASCADE,
+    follow_state community_follower_state NOT NULL,
+    PRIMARY KEY (multi_community_id, person_id)
```
The primary key for all the actions tables is person id first: 

`"community_actions_pkey" PRIMARY KEY, btree (person_id, community_id)`

dessalines replied:
NM, you fixed already.

dullbananas replied:
Before, there were indexes on `(multi_community_id, person_id)` and `(person_id)`. With just the primary key change, these indexes are now `(person_id, multi_community_id)` and `(person_id)`, which is redundant and is not good enough for queries that filter by `multi_community_id`. What's needed is `(person_id, multi_community_id)` and `(multi_community_id)`.

Nutomic replied:
Okay added `CREATE INDEX idx_multi_community_follow_multi_id ON multi_community_follow (multi_community_id);`

dessalines commented on crates/routes/src/utils/setup_local_site.rs:
```diff
@@ -79,9 +80,25 @@ pub async fn setup_local_site(pool: &mut DbPool<'_>, settings: &Settings) -> Lem
   let site = Site::create(pool, &site_form).await?;
 
+  // create multi-comm follower account
+  let r: String = rand::rng()
```
It'd probably be easier to take the first few chars of a uuid, which we already have as a dep.

dullbananas commented on migrations/2025-06-04-103308_multi-community/up.sql:
```diff
@@ -75,11 +75,17 @@ WHERE
+CREATE INDEX idx_multi_community_follow_multi_id ON multi_community_follow (multi_community_id);
+
 CREATE INDEX idx_multi_followed_by ON multi_community_follow (person_id);
 
```
```suggestion
```

This index on `(person_id)` is only a waste of resources, so it should be removed.

The new primary key automatically creates an index on `(person_id, multi_community_id)`, which completely removes the need for the `(person_id)` index. Both indexes have `person_id` all the way to the left side of the column list, so a query with a filter on `person_id` can use either index to avoid scanning rows that don't have a matching `person_id` value.

dessalines reviewed (commented):
Nice, this is (I think) the most requested feature, so it'll be great to have.

The only thing I potentially see missing, is searching for these (adding them as a searchable type), but that can get added in another PR. I'll make an issue for it.

dessalines reviewed (approved):
This is big enough that I think others should take a look also before we merge.

cc @SleeplessOne1917 @dullbananas  @phiresky

dessalines commented:
> One thing still missing compared to other types is a MultiCommunityView. Is it worth adding that?

Hrm... I didn't think about that. It belongs to a person and an instance, so it would be useful to have those come back in the API responses. Simple inner_joins work.

dessalines commented on crates/db_views/community/src/api.rs:
```diff
@@ -336,25 +337,30 @@ pub struct ListMultiCommunities {
+#[cfg_attr(feature = "ts-rs", ts(optional_fields, export))]
 pub struct GetMultiCommunityResponse {
-  pub multi: MultiCommunity,
+  #[serde(flatten)]
+  pub multi: MultiCommunityView,
```
Don't flatten this, it isn't the standard for any of our views. I'd also probably use the verbose `multi_community_view`, just to be explicit, and it matches the other names.

Nutomic replied:
Without this the api tests are full of lines like `alphaMulti.multi.multi.name` and `alphaMulti.multi.owner.id` which is very ugly.

dessalines replied:
Usually we use `res` to differentiate them from the views and objectsn within. So:

`alphaMultiRes.multi_community_view.multi_community.id`

Its verbose but clear as to what each sub object is.

dessalines commented on api_tests/run-federation-test.sh:
```diff
@@ -11,7 +11,7 @@ killall -s1 lemmy_server || true
 
 pnpm i
-pnpm api-test || true
+pnpm api-test-community || true
```
Needs changed back to test all.

dessalines commented:
Besides the above, and merge conflicts, it looks good. Feel free to merge once that's addressed and tests are passing.
